atc/configwarning.go

```diff
@@ -3,21 +3,44 @@ package atc
 import (
 	"fmt"
 	"regexp"
+	"strings"
 )
 
 type ConfigWarning struct {
 	Type    string `json:"type"`
 	Message string `json:"message"`
 }
 
+type InvalidIdentifierError struct {
+	Identifier string
+	Context    []string
+}
+
+func (e *InvalidIdentifierError) Error() string {
+	return fmt.Sprintf("'%s' is not a valid identifier", e.Identifier)
+}
+
+func (e *InvalidIdentifierError) ConfigWarning() ConfigWarning {
+	warning := ConfigWarning{
+		Type:    "invalid_identifier",
+		Message: e.Error(),
+	}
+
+	if e.Context != nil {
+		warning.Message = fmt.Sprintf("%s: %s", strings.Join(e.Context, ""), e.Error())
+	}
+
+	return warning
+}
+
 var validIdentifiers = regexp.MustCompile(`^\p{L}[\p{L}\d\-.]*$`)
 
 func ValidateIdentifier(identifier string, context ...string) error {
```

Comment:
Does it make sense for this to just return a `*ConfigWarning` instead of an `error`? That way, we won't need to use `errors.As` to ensure the error type is correct everywhere?

Reply:
It does, but I thought it would ease the transition from warning to errors in the next phase. I see those error type checks as temporary for now. Let me know if you'd rather have it return `*ConfigWarning`, it would make it cleaner for now...

Reply:
I think I'd prefer if it returned `*ConfigWarning`, since then when we change it to `error` we'll get compiler errors wherever it's called. Right now, if `ValidateIdentifier` were to return an error that was not an `*InvalidIdentifierError`, we just ignore it (since `errors.As` would return `false`). But if we get compiler errors, we're forced to think about how we handle it (and there aren't that many places we'd need to change, so it shouldn't be too much effort!)